### PR TITLE
Refactor Yum finders to use a common base class, add RHEL finder

### PR DESCRIPTION
**Pull request by 0xDEC0DE**
Oh boy, this one is a **MASSIVE** amount of changes, and will likely merit much discussion, as many of the decisions that went into this were made unilaterally and with zero input from anyone.

### Create a new common `YumFinder` class

This creates a new, common finder for all Yum-based distros.  The "fast lookup" patterns have been completely discarded in favour of "fuzzy" repomd lookups, with verification via HTTP `HEAD` requests.

These finders must be initialized with a canonical search list of repository URLs containing `repodata/repomd.xml` files.  The intention is for sub-classes of this finder to provide helper methods to fill out sensible default search directories, while also allowing for users to provide their own lists (e.g., for an internal-only mirror of packages)

### Refactor Yum-based finders as a subclasses of `YumFinder`

All of the previously-custom XPath searching code is now relegated to building search lists to feed the initializer in a default configuration.  This makes such things far more palatable, since by design SouFi will only try to make sense out of HTML tag soup when talking to well-known providers (i.e., not local mirrors with god-knows-what index modes enabled)

### Adds functional tests

A very (**VERY**) small subset of functionality is under test from this, mostly to demonstrate real-world improvements where these finders can locate sources the previous versions cannot.  They are disabled by default.

Fixes Issue #9 
Fixes Issue #10 
Fixes Issue #11

### CODE REVIEW

**Comment by 0xDEC0DE** [comment]
Please do not squash+merge.

File: soufi/finders/yum.py
Comment (0xDEC0DE) [inline soufi/finders/yum.py:34]: I realize this may not be the correct behavior, but this is a fully-capable finder on its own, provided that it is initialized with a valid set of search paths.  It's just that we're nominally discouraging that in favor of loading the derived finders.

It **DOES** seems to work like this, though.  Suggestions for improvement welcome.

File: soufi/tests/finders/test_yum_finder.py
Comment (0xDEC0DE) [inline soufi/tests/finders/test_yum_finder.py:313]: Well, this is embarrassing...

But what's in a name, anyway? 😆

File: soufi/finders/rhel.py
Comment (0xDEC0DE) [inline soufi/finders/rhel.py:35]: I figure of anything in this PR, this particular decision to "punt" source lookups for UBI has cause to be one of the more contentious, so I'll attempt to get in front of that be elaborating on what I have in the comments here:

As one can see from the provided list, Red Hat is fond of completely restructuring their repository layouts between major releases.  There is effectively zero commonality between UBI 7 and UBI 8, so it stands to reason that we will have absolutely no idea how to look for repos in UBI 9 when it ships.  This would infer that:

1. To enable built-in support for UBI 9 in the future, a new release of SouFi will be required regardless
2. To facilitate support for UBI 7+8+9 with this code, one would simply need to copy+paste the default search list, and the UBI 9 repo dirs, and initialize the finder with the whole list.

Given the above, I think the "bare minimum effort" approach taken here is justified.  It's also really, really fast. 😆

**Comment by juledwar** [comment]
Holy massive PR batman. I'll need some time to digest and will probably not merge until you're around again.

File: soufi/finders/yum.py
Comment (juledwar) [inline soufi/finders/yum.py:139]: Surely there's a library that does this? There's a bunch of stuff in the yum python module that makes me think we're missing out a bit.

File: soufi/finders/yum.py
Comment (juledwar) [inline soufi/finders/yum.py:158]: Kinda surprised you went this route, with your regex smarts.

File: soufi/finders/yum.py
Comment (juledwar) [inline soufi/finders/yum.py:162]: Wrap it in another method that doesn't catch the timeout!

File: soufi/__init__.py
Comment (juledwar) [inline soufi/__init__.py:247]: I don't understand what you're trying to do here. Why would 'optimal' be in the repo string?

**Review by juledwar — CHANGES_REQUESTED** [review]
I've reviewed all the main code so far, but not looked at the tests. Just dropping this down now and will come back for more tomorrow.

File: soufi/finders/yum.py
Comment (0xDEC0DE) [reply soufi/finders/yum.py:139]: There is, but I recall it having some issues, so I stole the useful bits and reimplemented it how I liked it

File: soufi/finders/yum.py
Comment (0xDEC0DE) [reply soufi/finders/yum.py:158]: Now you have two problems, etc. 😆

File: soufi/__init__.py
Comment (0xDEC0DE) [reply soufi/__init__.py:247]: ...because that's how you wrote it?  You offered `--repo optimal` as a shortcut in the CLI.

We *could* throw all that stuff down the stairs, but I was aiming to preserve existing functionality.

File: soufi/finders/yum.py
Comment (0xDEC0DE) [reply soufi/finders/yum.py:113]: Yes, which is why I mentioned one in the comment above. I also have a few explicit examples in the functional tests I wrote.

File: soufi/finders/centos.py
Comment (0xDEC0DE) [reply soufi/finders/centos.py:16]: Hahaha, I only did it this way to go along with precedent. Intensely happy to change it.

File: soufi/__init__.py
Comment (0xDEC0DE) [reply soufi/__init__.py:162]: Cry more. :trollface:

File: soufi/finders/centos.py
Comment (0xDEC0DE) [reply soufi/finders/centos.py:10]: Oh... The comment got separated from the import it was referring to, which clearly confused me.  Easy enough to put it back.

File: soufi/finders/centos.py
Comment (juledwar) [reply soufi/finders/centos.py:16]: > Hahaha, I only did it this way to go along with precedent. Intensely happy to change it.

I don't even know why I did it like that :sob:

File: soufi/finders/yum.py
Comment (juledwar) [reply soufi/finders/yum.py:113]: > Yes, which is why I mentioned one in the comment above. I also have a few explicit examples in the functional tests I wrote.

Cool, let's do eeeet.

File: soufi/finders/yum.py
Comment (juledwar) [reply soufi/finders/yum.py:139]: Ok that needs justifying in a comment please. Otherwise we'll come back here in a year and think "why didn't this idiot just use X library" ...

File: soufi/__init__.py
Comment (juledwar) [reply soufi/__init__.py:247]: Oh, drat, sorry, I completely forgot I had done that.

**Comment by 0xDEC0DE** [comment]
That last force-push was ill-advised, in the future I'll do individual commits, and I'll rewrite history on the branch as needed before merging.

Here's the diff:

```
diff --git a/soufi/finder.py b/soufi/finder.py
index 230959e..7099606 100644
--- a/soufi/finder.py
+++ b/soufi/finder.py
@@ -226,7 +226,11 @@ class FinderFactory:
         for _, module, _ in pkgutil.iter_modules(soufi.finders.__path__):
             mod = importlib.import_module(f"soufi.finders.{module}")
             for _name, obj in mod.__dict__.items():
-                if isclass(obj) and issubclass(obj, SourceFinder):
+                if (
+                    isclass(obj)
+                    and issubclass(obj, SourceFinder)
+                    and not hasattr(obj.distro, '__isabstractmethod__')
+                ):
                     # Add class object to our dict.
                     self._finders[obj.distro] = obj
 
@@ -238,7 +242,7 @@ class FinderFactory:
     @property
     def supported_types(self):
         """Return a list of known Finder types."""
-        return [key for key in self._finders.keys() if key is not None]
+        return list(self._finders.keys())
 
 
 factory = FinderFactory()
diff --git a/soufi/finders/yum.py b/soufi/finders/yum.py
index e200a07..f73e5f0 100644
--- a/soufi/finders/yum.py
+++ b/soufi/finders/yum.py
@@ -1,6 +1,7 @@
 # Copyright (c) 2021 Cisco Systems, Inc. and its affiliates
 # All rights reserved.
 
+import abc
 import functools
 import urllib
 
@@ -12,8 +13,8 @@ from soufi import exceptions, finder
 TIMEOUT = 30  # seconds
 
 
-class YumFinder(finder.SourceFinder):
-    """A base class for Yum-based repositories.
+class YumFinder(finder.SourceFinder, metaclass=abc.ABCMeta):
+    """An abstract base class for making Yum-based finders.
 
     Subclasses of this should provide methods for setting up the repository
     search lists to pass to __init__.
@@ -31,13 +32,19 @@ class YumFinder(finder.SourceFinder):
            to lookup that package in the source repos.
     """
 
-    distro = None
-
     def __init__(self, *args, source_repos, binary_repos, **kwargs):
         self.source_repos = source_repos
         self.binary_repos = binary_repos
         super().__init__(*args, **kwargs)
 
+    @abc.abstractmethod
+    def _get_source_repos(self):
+        raise NotImplementedError  # pragma: nocover
+
+    @abc.abstractmethod
+    def _get_binary_repos(self):
+        raise NotImplementedError  # pragma: nocover
+
     def _find(self):
         source_url = self.get_source_url()
         return YumDiscoveredSource([source_url])
diff --git a/soufi/tests/finders/test_yum_finder.py b/soufi/tests/finders/test_yum_finder.py
index eef97d6..0f75c89 100644
--- a/soufi/tests/finders/test_yum_finder.py
+++ b/soufi/tests/finders/test_yum_finder.py
@@ -15,6 +15,18 @@ from soufi.finders import yum
 from soufi.testing import base
 
 
+class YumFinderImpl(yum.YumFinder):
+    """An implementation for testing the YumFimder ABC."""
+
+    distro = 'yum'
+
+    def _get_source_repos(self):
+        pass
+
+    def _get_binary_repos(self):
+        pass
+
+
 class BaseYumTest(base.TestCase):
     class FakePackage:
         vr = None
@@ -46,7 +58,7 @@ class BaseYumTest(base.TestCase):
             kwargs['source_repos'] = []
         if 'binary_repos' not in kwargs:
             kwargs['binary_repos'] = []
-        return yum.YumFinder(name, version, SourceType.os, **kwargs)
+        return YumFinderImpl(name, version, SourceType.os, **kwargs)
 
     def make_response(self, data, code):
         fake_response = mock.MagicMock()
```

```
diff --git a/soufi/finders/rhel.py b/soufi/finders/rhel.py
index 1996363..753bf14 100644
--- a/soufi/finders/rhel.py
+++ b/soufi/finders/rhel.py
@@ -42,17 +42,21 @@ class RHELFinder(yum_finder.YumFinder):
         **kwargs,
     ):
         if not source_repos:
-            source_repos = [
-                f"{DEFAULT_REPO}{dir}source/SRPMS/"
-                for dir in self.default_search_dirs
-            ]
+            source_repos = self._get_source_repos()
         if not binary_repos:
-            binary_repos = [
-                f"{DEFAULT_REPO}{dir}os/" for dir in self.default_search_dirs
-            ]
+            binary_repos = self._get_binary_repos()
         super().__init__(
             *args,
             source_repos=source_repos,
             binary_repos=binary_repos,
             **kwargs,
         )
+
+    def _get_source_repos(self):
+        return [
+            f"{DEFAULT_REPO}{dir}source/SRPMS/"
+            for dir in self.default_search_dirs
+        ]
+
+    def _get_binary_repos(self):
+        return [f"{DEFAULT_REPO}{dir}os/" for dir in self.default_search_dirs]
```

File: soufi/finders/rhel.py
Comment (juledwar) [inline soufi/finders/rhel.py:47]: Can this just not go in the base class? Or am I missing something?

Comment (0xDEC0DE) [reply soufi/finders/rhel.py:47]: It could, and for a minute it did, but it made things look really awkward to my eyes, so I kept it like this for clarity.

Comment (juledwar) [reply soufi/finders/rhel.py:47]: It looks weirder like this to my eyes... YMMV :)

**Comment by juledwar** [comment]
Looks ok so far but there's a coverage error.

**Comment by 0xDEC0DE** [comment]
> Looks ok so far but there's a coverage error.

When I flopped the trailing-slash logic in the Photon finder, it became necessary to also flop the tests, and I didn't.  Whoops.

But looking at it again, I went with `str.endswith`and slicing in Photon, whereas elsewhere I just used `str.rstrip`.  I'll push an update to "harmonize" that.

File: soufi/finders/yum.py
Comment (0xDEC0DE) [reply soufi/finders/yum.py]: Though I am loathe to cache `stat(2)` calls for all of the obvious reasons, recent studies have shown that HTTP HEAD requests are, in fact, **NOT** `stat(2) calls!  😆 

An update to this to cache HTTP HEAD calls has been pushed.

File: soufi/finders/yum.py
Comment (juledwar) [inline soufi/finders/yum.py:144]: Add a ticket in GH and reference here please.

File: soufi/finders/rhel.py
Comment (0xDEC0DE) [reply soufi/finders/rhel.py:47]: After thinking about it a bit more, it does seem rather bonkers to define an ABC that demands a pair of methods to be defined in subclasses, and then never calls those methods itself.

So I'll get over my aesthetic concerns and DRY this out, as it does seem to make the most sense.

File: soufi/finders/photon.py
Comment (0xDEC0DE) [reply soufi/finders/photon.py]: I decided instead to do a search-and-destroy of everything doing this, and instead do the "fixup" only in the `YumFinder._get_repo` method (which is cached)

I've been burned a couple of times by trying to use `os.path` functions to manipulate URLs, and have developed a healthy superstition against it, even when I know it will be safe.

File: soufi/finders/rhel.py
Comment (juledwar) [inline soufi/finders/rhel.py:42]: ART!

File: soufi/finders/yum.py
Comment (0xDEC0DE) [reply soufi/finders/yum.py:115]: Different name **from the source package**.  I'll push a clarification.

File: soufi/finders/yum.py
Comment (0xDEC0DE) [reply soufi/finders/yum.py:81]: As you wish.

Comment (0xDEC0DE) [reply soufi/finders/yum.py:81]: It originally was (well, it originally always looked up `self.version` unconditionally), but I busted it down to something that could be overridden for the Photon use-case, where it comes in very handy.

I'm more partial to achieving consistency by going in the other direction and having all parameters be optional, with the defaults being `self`.  But I think this half-and-half approach

File: soufi/finders/centos.py
Comment (juledwar) [reply soufi/finders/centos.py:83]: You've pretty much written up my problem with it so I guess I didn't make enough sense to start with.

My beef is with the HEAD calls that will never get used. Iterating that list of locations "to completion" is wasting time and resources and potentially annoying the server operators with the repeated HEAD calls.

If we're searching for 1000s of packages, the cache will still "warm up" naturally without a ton of HEAD calls. It's just that you'd not be doing _unnecessary_ HEAD calls, by doing the actual GET instead and moving on if it fails.

**Comment by 0xDEC0DE** [comment]
Not exactly rigorous, since I only ran each once, but it's still a pretty stark difference:

## CentOS Discovery as List:
```
real	3m35.277s
user	0m40.642s
sys	0m8.056s
```

## CentOS Discovery as Generator:
```
real	1m55.055s
user	0m39.000s
sys	0m8.508s
```

File: soufi/finders/yum.py
Comment (juledwar) [inline soufi/finders/yum.py:51]: These should be public (ie no _) by definition - they're abstract.

File: soufi/finders/yum.py
Comment (juledwar) [inline soufi/finders/yum.py:38]: It feels like a lot of complexity can be removed if you get rid of the source_repos and binary_repos init args here, and rely solely on the abstract get_X_repos() methods to provide the list/generator when requested. You can them remove the stuff in the derived finder that is attempting to re-initialise the generator.

File: soufi/finders/yum.py
Comment (juledwar) [inline soufi/finders/yum.py:51]: Make these public.

File: soufi/finders/centos.py
Comment (juledwar) [inline soufi/finders/centos.py:70]: Yeah I think most of this can disappear once the init is simplified.

File: soufi/finders/yum.py
Comment (juledwar) [inline soufi/finders/yum.py:180]: Make this public, it's used by derived classes. Also, a classmethod is appropriate since you're calling other functions defined in the class.

File: soufi/finders/yum.py
Comment (juledwar) [inline soufi/finders/yum.py:190]: Classmethod

File: soufi/finders/yum.py
Comment (juledwar) [inline soufi/finders/yum.py:197]: Then it should be public.

File: soufi/finders/photon.py
Comment (juledwar) [inline soufi/finders/photon.py:54]: Looks like the walk* funcs need to be public too.

**Review by juledwar — CHANGES_REQUESTED** [review]

File: soufi/tests/finders/test_centos_finder.py
Comment (juledwar) [inline soufi/tests/finders/test_centos_finder.py:74]: This is incredibly unreadable. What about
`self.assertThat(result, SameMembers(['1.0.123', '2.1.3456', '3.7.89']))`

**Comment by juledwar** [comment]
I've run out of steam looking at this, there's enough feedback for now.

File: soufi/finders/yum.py
Comment (0xDEC0DE) [reply soufi/finders/yum.py:51]: I went back and forth on this one for a bit:

Are they public, since they NEED to be implemented by subclasses of `YumFinder`?

Or are they private, since they are implementation details of the main `SourceFinder` class, which only exposes one public method, `find()`, and these methods should never be called directly?  

In the end, I let the fact that you created an abstract private method called `SourceFinder._find()` decide the issue for me.  So let me drink in the sweet, sweet irony of you pushing back on this.  Ahhhh...  😆

File: soufi/finders/yum.py
Comment (0xDEC0DE) [reply soufi/finders/yum.py:38]: If I'm understanding this ask correctly (hint: it's early here so likely I'm not) that would remove the ability to instantiate the finder using private/local/etc. mirrors, which is absolutely something that I want to be able to do.

File: soufi/finders/yum.py
Comment (0xDEC0DE) [reply soufi/finders/yum.py:190]: OK, but it's really...  not.

File: soufi/tests/finders/test_centos_finder.py
Comment (0xDEC0DE) [reply soufi/tests/finders/test_centos_finder.py:74]: That's fair.  The intent was to demonstrate that the results were coming back in a certain order, and skipping the bogus one.  Something clearly went wrong with the execution.

File: soufi/finders/centos.py
Comment (0xDEC0DE) [reply soufi/finders/centos.py:122]: It does need to be a closure!  `yield from` demands it, for reasons.

File: soufi/finders/photon.py
Comment (0xDEC0DE) [reply soufi/finders/photon.py:54]: Not really — they're definitely implementation details of the `YumFinder` and aren't supposed to be called or cared about by subclasses.  If you're overriding them, something has gone horribly wrong (e.g., Photon)

File: soufi/finders/centos.py
Comment (juledwar) [reply soufi/finders/centos.py:122]: ..blink.. WTF. OK then.

File: soufi/finders/yum.py
Comment (juledwar) [reply soufi/finders/yum.py]: You can still wrap that inside the get method I think. Just take it out of the base class. It's the direct cause of the confusion and need to reset the generator with a flag.

Which brings me to another thought - if someone passes a private list then it won't act as a generator at all, so I think this whole area needs some redesigning to wrap those in generators.

File: soufi/finders/yum.py
Comment (juledwar) [reply soufi/finders/yum.py]: I figured you'd mention that :smile: I screwed that one up, to be fair, and it should be renamed to something else entirely. I used `_find` because `find` is already used.

> Are they public, since they NEED to be implemented by subclasses

Yes, that is the definition of public. This is not C++! We don't do "protected".

File: soufi/finders/yum.py
Comment (juledwar) [reply soufi/finders/yum.py]: The fact that it's getting called _from within the same class_ using__ `YumFinder._head_url` means it should be a classmethod.

File: soufi/finders/photon.py
Comment (juledwar) [reply soufi/finders/photon.py:54]: If something has gone horribly wrong then the design is wrong. Is there any aspect you can factor out and allow that to be overridden?

File: soufi/finders/yum.py
Comment (0xDEC0DE) [reply soufi/finders/yum.py]: I *just** moved this logic into the base class to get rid of all the lookalike `__init__` methods in all the derived classes, now you want me to put all that back?

File: soufi/tests/finders/test_centos_finder.py
Comment (0xDEC0DE) [reply soufi/tests/finders/test_centos_finder.py]: OK, so this one is REALLY esoteric, but:

I am still dealing with this regression: https://youtrack.jetbrains.com/issue/PY-48718

...which means my test runner swallows most output errors and simply prints "OK" and "Failure" when running.  So it really helps me to bind damn-near everything in tests to an lvalue so I can set a breakpoint and look at the test results in a debugger when tests fail.  So while I can write comparisons with `testtools.matchers.SameMembers`, and it will just use the generator and Do The Right Thing, this test pattern is far more useful to me.

File: soufi/finders/yum.py
Comment (bigjools) [reply soufi/finders/yum.py]: That was before you added the generator! You can still wrap everything in a generator in this __init__ though if you want.  I just dislike the flag that says to restart the generator, it seems unnecessary if you just make everything a generator.

File: soufi/tests/finders/test_centos_finder.py
Comment (bigjools) [reply soufi/tests/finders/test_centos_finder.py]: Ok no worries. I hate tools sometimes.

> It's been over 6 months, and I ostensibly pay for support.

Ouch :laughing:

**Review by juledwar — APPROVED** [review]
It works. I will follow up later with changes to incrementally improve instead of holding up this now. Great work, really!

**Pull request merged**
